core/src/apps/base.py

```diff
@@ -0,0 +1,189 @@
+    from trezor.messages.Ping import Ping
+
+
+def get_features() -> Features:
+    f = Features()
+    f.vendor = "trezor.io"
+    f.language = "en-US"
+    f.major_version = utils.VERSION_MAJOR
+    f.minor_version = utils.VERSION_MINOR
+    f.patch_version = utils.VERSION_PATCH
+    f.revision = utils.GITREV.encode()
+    f.model = utils.MODEL
+    f.device_id = storage.device.get_device_id()
+    f.label = storage.device.get_label()
+    f.pin_protection = config.has_pin()
+    f.pin_cached = config.is_unlocked()
+    f.passphrase_protection = storage.device.is_passphrase_enabled()
+
+    if utils.BITCOIN_ONLY:
+        f.capabilities = [
+            Capability.Bitcoin,
+            Capability.Crypto,
+            Capability.Shamir,
+            Capability.ShamirGroups,
+            Capability.PassphraseEntry,
+        ]
+    else:
+        f.capabilities = [
+            Capability.Bitcoin,
+            Capability.Bitcoin_like,
+            Capability.Binance,
+            Capability.Cardano,
+            Capability.Crypto,
+            Capability.EOS,
+            Capability.Ethereum,
+            Capability.Lisk,
+            Capability.Monero,
+            Capability.NEM,
+            Capability.Ripple,
+            Capability.Stellar,
+            Capability.Tezos,
+            Capability.U2F,
+            Capability.Shamir,
+            Capability.ShamirGroups,
+            Capability.PassphraseEntry,
+        ]
+    f.sd_card_present = sdcard.is_present()
+
+    # private fields:
+    if config.is_unlocked():
+        # While this is technically not private, we can't reliably find the value while
+        # locked. Instead of sending always False, we choose to not send it.
+        f.initialized = storage.is_initialized()
+
+        f.needs_backup = storage.device.needs_backup()
+        f.unfinished_backup = storage.device.unfinished_backup()
+        f.no_backup = storage.device.no_backup()
+        f.flags = storage.device.get_flags()
+        f.recovery_mode = storage.recovery.is_in_progress()
+        f.backup_type = mnemonic.get_type()
+        f.sd_protection = storage.sd_salt.is_enabled()
+        f.wipe_code_protection = config.has_wipe_code()
+        f.passphrase_always_on_device = storage.device.get_passphrase_always_on_device()
+
+    return f
+
+
+async def handle_Initialize(ctx: wire.Context, msg: Initialize) -> Features:
+    features = get_features()
+    if msg.session_id:
+        msg.session_id = bytes(msg.session_id)
+    features.session_id = cache.start_session(msg.session_id)
+    return features
+
+
+async def handle_GetFeatures(ctx: wire.Context, msg: GetFeatures) -> Features:
+    return get_features()
+
+
+async def handle_Cancel(ctx: wire.Context, msg: Cancel) -> NoReturn:
+    raise wire.ActionCancelled
+
+
+async def handle_LockDevice(ctx: wire.Context, msg: LockDevice) -> Success:
+    lock_device()
+    workflow.kill_default()  # this restarts the homescreen to show lock icon
```

Comment:
It seems to me you also need to kill all workflow tasks here.

Reply:
if you mean "abort any currently running action", that happens as part of the session handling: https://github.com/trezor/trezor-firmware/blob/matejcik/softlock/core/src/trezor/wire/__init__.py#L342-L350

that's how `handle_Cancel` is a no-op

if you mean something else (hard kill?), i'm not sure what exactly that is

Reply:
What basically mean is:
```
for task in workflow.tasks:
    loop.close(task)
```
It needs to be written differently though, because `loop.close()` modifies `workflow.tasks`.

Reply:
what do you hope to accomplish by this, that doesn't already happen implicitly?

Reply:
If I understand correctly, what you are trying to say is that workflows which are part of a session get closed automatically with that session. Well what about workflows that are not part of a session?

Reply:
fixed in 9b2ee88

Reply:
this is in fact fixed in #1027, via `loop.spawn` and its use in workflow